### PR TITLE
Refactor add_states in Bloch, fix style, add changelog

### PR DESCRIPTION
**Pull request by suryanshcode2601**
## Summary
Refactors `add_states` in the `Bloch` class to improve readability and
maintainability, aligned with the structure of `add_vectors` and `add_points`.

## Root cause of previous failures
`np.asarray(state)` was used to normalize input, but numpy cannot reliably
build a numeric array from a list of `Qobj` instances — it creates an object
array and all subsequent `ndim`/`shape` checks behave unexpectedly. Replaced
with an explicit `isinstance` check.

## Changes
- Replaced `np.asarray(state)` with `isinstance` check for safe input normalization
- Added early validation for `kind` parameter before any processing
- Added validation for `alpha` parameter (must be between 0 and 1)
- Extracted color normalization into `_normalize_colors` helper method,
  consistent with the pattern used in `add_vectors` and `add_points`
- Simplified loop using `zip(state, colors)` — no index needed

## Testing
Ran `pytest qutip/tests/test_bloch.py -v` locally:
- 24 passed, 0 failed
- Verified with single state, list of states, invalid `kind`, invalid `alpha`,
  and mismatched colors

## Style
- Ran `pycodestyle qutip/bloch.py` — no violations

### CODE REVIEW

**Comment by veronikakurth** [comment]
@suryanshcode2601 this PR seems to be a duplicate of #2842 - if not, could you please describe how it is different from it and which points of the code review do you address here?

**Comment by suryanshcode2601** [comment]
Hi, thanks for the feedback! This PR (#2844) is not a duplicate — it is a direct replacement for #2842, submitted after addressing the review comments received there. I should have mentioned this more clearly upfront, and I apologize for the confusion.

Here is how #2844 differs from and improves upon #2842:

1. Fixed the root bug that was missed in #2842
The original code used `np.asarray(state)` to normalize input, but numpy cannot reliably build a numeric array from a list of `Qobj` instances — it silently creates an object array, causing all subsequent `ndim`/`shape` checks to behave incorrectly. This PR replaces that with an explicit `isinstance` check, which is the correct approach.

2. Addressed the "make a proper attempt at refactoring" feedback
The `add_states` method previously mixed input validation, color handling, and dispatch logic all in one place. This PR separates those concerns by extracting color normalization into a dedicated `_normalize_colors` helper method, consistent with the pattern already used in `add_vectors` and `add_points`.

3. Tests were actually run this time
I ran `pytest qutip/tests/test_bloch.py -v` locally — 24 passed, 0 failed. I also verified edge cases: single state, list of states, invalid `kind`, invalid `alpha`, and mismatched colors.

4. Style was checked
Ran `pycodestyle qutip/bloch.py` — no violations.

5. Improved PR description
The description in this PR is structured with sections (Root cause, Changes, Testing, Style) to be informative and reviewable, unlike the sparse description in #2842.

I'm happy to close #2842 formally if that helps keep things clean. Please let me know if there's anything else to address!

**Comment by veronikakurth** [comment]
> Hi, thanks for the feedback! This PR (#2844) is not a duplicate — it is a direct replacement for #2842, submitted after addressing the review comments received there. I should have mentioned this more clearly upfront, and I apologize for the confusion.
> 
> Here is how #2844 differs from and improves upon #2842:
> 
> 1. Fixed the root bug that was missed in [Refactor add_states in Bloch for improved readability and structure #2842](https://github.com/qutip/qutip/pull/2842)
>    The original code used `np.asarray(state)` to normalize input, but numpy cannot reliably build a numeric array from a list of `Qobj` instances — it silently creates an object array, causing all subsequent `ndim`/`shape` checks to behave incorrectly. This PR replaces that with an explicit `isinstance` check, which is the correct approach.
> 2. Addressed the "make a proper attempt at refactoring" feedback
>    The `add_states` method previously mixed input validation, color handling, and dispatch logic all in one place. This PR separates those concerns by extracting color normalization into a dedicated `_normalize_colors` helper method, consistent with the pattern already used in `add_vectors` and `add_points`.
> 3. Tests were actually run this time
>    I ran `pytest qutip/tests/test_bloch.py -v` locally — 24 passed, 0 failed. I also verified edge cases: single state, list of states, invalid `kind`, invalid `alpha`, and mismatched colors.
> 4. Style was checked
>    Ran `pycodestyle qutip/bloch.py` — no violations.
> 5. Improved PR description
>    The description in this PR is structured with sections (Root cause, Changes, Testing, Style) to be informative and reviewable, unlike the sparse description in [Refactor add_states in Bloch for improved readability and structure #2842](https://github.com/qutip/qutip/pull/2842).
> 
> I'm happy to close #2842 formally if that helps keep things clean. Please let me know if there's anything else to address!



> Hi, thanks for the feedback! This PR (#2844) is not a duplicate — it is a direct replacement for #2842, submitted after addressing the review comments received there. I should have mentioned this more clearly upfront, and I apologize for the confusion.
> 
> Here is how #2844 differs from and improves upon #2842:
> 
> 1. Fixed the root bug that was missed in [Refactor add_states in Bloch for improved readability and structure #2842](https://github.com/qutip/qutip/pull/2842)
>    The original code used `np.asarray(state)` to normalize input, but numpy cannot reliably build a numeric array from a list of `Qobj` instances — it silently creates an object array, causing all subsequent `ndim`/`shape` checks to behave incorrectly. This PR replaces that with an explicit `isinstance` check, which is the correct approach.
> 2. Addressed the "make a proper attempt at refactoring" feedback
>    The `add_states` method previously mixed input validation, color handling, and dispatch logic all in one place. This PR separates those concerns by extracting color normalization into a dedicated `_normalize_colors` helper method, consistent with the pattern already used in `add_vectors` and `add_points`.
> 3. Tests were actually run this time
>    I ran `pytest qutip/tests/test_bloch.py -v` locally — 24 passed, 0 failed. I also verified edge cases: single state, list of states, invalid `kind`, invalid `alpha`, and mismatched colors.
> 4. Style was checked
>    Ran `pycodestyle qutip/bloch.py` — no violations.
> 5. Improved PR description
>    The description in this PR is structured with sections (Root cause, Changes, Testing, Style) to be informative and reviewable, unlike the sparse description in [Refactor add_states in Bloch for improved readability and structure #2842](https://github.com/qutip/qutip/pull/2842).
> 
> I'm happy to close #2842 formally if that helps keep things clean. Please let me know if there's anything else to address!

Thanks for the feedback and yes, please feel free to close the previous pull request as there is no need to keep both open.

**Comment by suryanshcode2601** [comment]
Hey @veronikakurth , I wanted to know what else is required for me to do so that this PR can be merged

File: doc/changes/2844.trivial [hunk 1]
```diff
@@ -0,0 +1 @@
+Refactored ``add_states`` in ``Bloch`` to improve readability, add input validation, and extract color normalization into a helper method.
```
Comment (veronikakurth) [inline doc/changes/2844.trivial:1]: What is tag "trivial" for this changelog?

File: qutip/bloch.py [hunk 2]
```diff
@@ -390,49 +390,83 @@ def add_states(self, state: Qobj,
...
             Transparency value for the vectors. Values between 0 and 1.
         """
-        state = np.asarray(state)
+        if not isinstance(state, (list, np.ndarray)):
+            state = [state]
 
-        if state.ndim == 0:
-            state = state[np.newaxis]
+        # Validate all states are Qobj
+        for st in state:
```
Comment (veronikakurth) [inline qutip/bloch.py:397]: The type validation is already performed by the `expect` function called in the `_state_to_cartesian_coordinates` function. 
Hence is the question: What kind of benefit does this additional validation give?

File: qutip/bloch.py [hunk 3]
```diff
@@ -390,49 +390,83 @@ def add_states(self, state: Qobj,
         alpha : float, default=1.
             Transparency value for the vectors. Values between 0 and 1.
         """
-        state = np.asarray(state)
+        if not isinstance(state, (list, np.ndarray)):
+            state = [state]
```
Comment (veronikakurth) [inline qutip/bloch.py:394]: With this clause, you introduce an ambiguity on whether variable `state` will be a Python list or a NumPy array. If there is no specific reason to have it, I'd suggest to leave `state = np.asarray(state)` as it is.

File: qutip/bloch.py (hunk 2 above)
Comment (veronikakurth) [reply qutip/bloch.py:397]: To prove why it's useful, feel free to add a corresponding test. 
Namely, to test for behavior when states are of incorrect types - this might be lacking in the current test cases.

File: qutip/bloch.py [hunk 4]
```diff
@@ -390,49 +390,83 @@ def add_states(self, state: Qobj,
...
+                raise ValueError(
+                    "The included states are not valid. "
+                    "State should be a Qobj or a list of Qobj."
+                )
 
-        if state.ndim != 1:
-            raise ValueError("The included states are not valid. "
-                             "State should be a Qobj or a list of Qobj.")
+        # Validate kind early, before doing any work
+        if kind not in ('vector', 'point'):
```
Comment (veronikakurth) [inline qutip/bloch.py:405]: Makes sense that the validation happens at the very beginning of the function's body, I agree.

File: qutip/bloch.py [hunk 5]
```diff
@@ -390,49 +390,83 @@ def add_states(self, state: Qobj,
...
+            raise ValueError(
+                "The included kind is not valid. "
+                f"It should be vector or point, not {kind}."
+            )
 
-        if colors is not None:
-            colors = np.asarray(colors)
+        # Validate alpha
+        if not (0.0 <= alpha <= 1.0):
```
Comment (veronikakurth) [inline qutip/bloch.py:412]: I believe this check is redundant and the error message doesn't provide more information than `matplotlib` already does for us. Here, `alpha` is checked for type and value: https://github.com/matplotlib/matplotlib/blob/v3.10.8/lib/matplotlib/artist.py#L1013

File: qutip/bloch.py [hunk 6]
```diff
@@ -390,49 +390,83 @@ def add_states(self, state: Qobj,
...
+        n = len(state)
+        colors = self._normalize_colors(colors, n)
 
-            elif (
-                colors.ndim == 1
-                and (np.issubdtype(colors.dtype, np.floating)
-                     or np.issubdtype(colors.dtype, np.integer))):
-                colors = colors[np.newaxis]
-                colors = np.repeat(colors, [state.shape[0]], axis=0)
+        for st, color in zip(state, colors):
```
Comment (veronikakurth) [inline qutip/bloch.py:421]: Using `zip` or `enumerate` seems to be a matter of taste here. Does it add any functional benefit over `enumerate`?

File: qutip/bloch.py [hunk 7]
```diff
@@ -390,49 +390,83 @@ def add_states(self, state: Qobj,
...
+                self.add_points(vec, colors=[color], alpha=alpha)
 
-            if (
-                colors.shape[0] != state.shape[0] or colors.ndim > 2
-                or colors.ndim == 2
-                and not (np.issubdtype(colors.dtype, np.floating)
-                         or np.issubdtype(colors.dtype, np.integer))):
-                raise ValueError("The included colors are not valid. "
-                                 "colors must have the same size as state.")
+    def _normalize_colors(self, colors, n: int):
```
Comment (veronikakurth) [inline qutip/bloch.py:428]: Makes sense to have it as a separate function to increase readability, I agree.

**Review by veronikakurth — CHANGES_REQUESTED** [review]
Thank you for your changes - please address the comments above.

File: qutip/bloch.py (hunk 3 above)
Comment (suryanshcode2601) [reply qutip/bloch.py:394]: You're right, I'll remove the isinstance branch. Since np.asarray on a list of Qobj creates an unreliable object array, I'll replace it entirely with an explicit isinstance check instead of using both together.

File: qutip/bloch.py (hunk 6 above)
Comment (suryanshcode2601) [reply qutip/bloch.py]: Using zip is intentional — since the index is never needed, zip is cleaner. I'll add a comment to make that clear.

**Comment by veronikakurth** [comment]
The review has not been fully addressed, and the PR has been abandoned by the author. Closing.
